### PR TITLE
Resolver construct to simplify promise/resolve/reject situations

### PR DESCRIPTION
**Pull request by dvoytenko**

### CODE REVIEW

File: src/promise.js [hunk 1]
```diff
@@ -44,3 +44,38 @@ export function all(promises) {
...
+    resolve = aResolve;
+    reject = aReject;
+  });
+  return new Resolver(promise, resolve, reject);
+}
+
+/**
+ * @template T
+ */
+class Resolver {
```
Comment (erwinmombay) [inline src/promise.js]: can we rename this to `Deferred` as its close to the jquery Deferred object. (where its an instance that has a .promise property and .resolve/.reject methods). i dont feel strongly either way so feel free to ignore.

Comment (dvoytenko) [inline src/promise.js]: But it's "Resolver" in Closure and YUL.

**Comment by ampsauce** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: b18ab6e6f7aebabf33a879c9e634697c833d92f8
Build details: https://travis-ci.org/ampsauce/amphtml/builds/99182606

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

File: src/promise.js (hunk 1 above)
Comment (erwinmombay) [inline src/promise.js]: sgtm.

File: src/promise.js [hunk 2]
```diff
@@ -44,3 +44,38 @@ export function all(promises) {
...
+  /**
+   * @param {!Promise<T>} promise
+   * @param {function(T)} resolve
+   * @param {function(*=)} reject
+   */
+  constructor(promise, resolve, reject) {
+    /** @const @type {!Promise<T>} */
+    this.promise = promise;
+    /** @const @type {function(T)} */
+    this.resolve = resolve.bind(null);
```
Comment (jridgewell) [inline src/promise.js]: Is there a reason to bind here? I don't know of a promise implementation that changes behavior due the to `this`.

File: src/promise.js (hunk 1 above)
Comment (jridgewell) [inline src/promise.js]: I agree that this should be Deferred, that's what all JS libraries call it.

**Comment by jridgewell** [comment]
Just note that Bluebird considers this an [anti-pattern](https://github.com/petkaantonov/bluebird/wiki/Promise-anti-patterns#the-deferred-anti-pattern).

**Comment by cramforce** [comment]
It's an anti pattern if overused. Promise's API not supporting this by
default is the right choice. But some cases are inherently better handled
with it. It is just important to never leak the deferred/resolver in a
public API and keep it constrained to a small scope.
Not a fan of Deferred. We are a bit "vorbelastet" because of
goog.async.Deferred :)

**Comment by dvoytenko** [comment]
The real antipattern is the leaking of resolver function itself. In some cases Resolver would make it better. That being said, I haven't yet bought into this idea either. I'll first take a look at how often we store the resolvers as instance vars. I did see recently an increase in those where we store pairs of promise/resolver. It's possible that those can be changed.

If I go ahead with this, I will put it into presubmit restriction set.

**Comment by dvoytenko** [comment]
So, I looked through most of cases we have today, and my opinion is that in most of cases, it's possible to refactor code to avoid resolver leaker. So, I'm gonna close this PR.

**Pull request closed without merging**